File: Problems/Algorithms/102. Binary Tree Level Order Traversal/level_order_traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        
        fn traversal(root: Option<Rc<RefCell<TreeNode>>>, res: &mut Vec<Vec<i32>>, level: usize) {
            match root {
                None => (),
                Some(r) => {
                    if res.len() == level {
                        res.push(Vec::new());
                    }
                    res[level].push(r.borrow().val);
                    traversal(r.borrow().left.clone(), res, level+1);
                    traversal(r.borrow().right.clone(), res, level+1);
                    ()
                }
            }
        }
        
        traversal(root.clone(), &mut res, 0);
        res
    }
}

File: Problems/Algorithms/102. Binary Tree Level Order Traversal/level_order_traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    result := make([][]int, 0)

    traversal(root, &result, 0)
    return result
}

func traversal(root *TreeNode, result *[][]int, level int) {
    if root == nil {
        return
    }

    if len(*result) == level {
        *result = append(*result, []int{})
    }
    (*result)[level] = append((*result)[level], root.Val)
    traversal(root.Left, result, level+1)
    traversal(root.Right, result, level+1)
}
